File: examples/haskellfun/haskellfun.hs
{-# LANGUAGE OverloadedStrings #-}

-- An example with higher-level Haskell functions. Haskell functions are
-- wrapped by hslua automatically for ensuring argument types and nubmers
-- and passing arguments from Lua stack to Haskell functions. Return values
-- are also handled by hslua automatically(so you don't put return value to
-- stack manually).

import qualified Data.ByteString as B
import Data.Monoid
import Foreign.Lua as Lua

main :: IO ()
main = runLua $ do
  openlibs
  registerhsfunction "concat" concat'
  registerhsfunction "pow" pow
  registerhsfunction "helloWorld" helloWorld
  loadfile "examples/haskellfun/haskellfun.lua"
  call 0 0

concat' :: B.ByteString -> B.ByteString -> IO B.ByteString
concat' s1 s2 = return $ s1 <> s2

pow :: Double -> Double -> IO Double
pow d1 d2 = return $ d1 ** d2

helloWorld :: IO B.ByteString
helloWorld = return "Hello, World!"

File: examples/haskellfun/haskellfun.lua
_HASKELLERR = {}

function catch_haskell(ret, err_msg)
    if ret == _HASKELLERR then
      print("Error caught from Haskell land: " .. err_msg)
      return
    end
    return ret
end

print(concat("hello", " world!"))
print(catch_haskell(pow(3.2, 5)))
print(catch_haskell(helloWorld()))
print(catch_haskell(pow("wrong")))
print(catch_haskell(pow(3)))
